RFD 0184: Automatic Updates for Teleport Agents

sclevine opened the pull request:
This RFD proposes a new mechanism for Teleport agents to automatically update to a version set by an operator via tctl.

Supersedes https://github.com/gravitational/teleport/pull/40190.

[Readable](https://github.com/gravitational/teleport/blob/rfd/0184-agent-auto-updates/rfd/0184-agent-auto-updates.md)

sclevine committed:
Create 0169-auto-updates-linux-agents.md

sclevine committed:
Fix github handle

sclevine committed:
Fix Github handle

sclevine committed:
Clarify jitter flag

sclevine committed:
Remove time question

sclevine committed:
Update rfd/0169-auto-updates-linux-agents.md

Co-authored-by: Russell Jones <russjones@users.noreply.github.com>

sclevine committed:
Update rfd/0169-auto-updates-linux-agents.md

Co-authored-by: Russell Jones <russjones@users.noreply.github.com>

sclevine committed:
Update rfd/0169-auto-updates-linux-agents.md

Co-authored-by: Russell Jones <russjones@users.noreply.github.com>

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
add editions

sclevine committed:
Installers and docs

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Downgrades

sclevine committed:
Feedback

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Remove last working copy of teleport

sclevine committed:
add step to ensure free disk space

sclevine committed:
Typos

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
feedback

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
apt purge

sclevine committed:
Only enable auto-upgrades if successful

sclevine committed:
reentrant lock

sclevine committed:
reset

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
add note on backups

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
Clarify restore/rollback process and validations

sclevine committed:
Added section on logging

sclevine committed:
Add schedules

sclevine committed:
immediate schedule + note on cycles and chains

sclevine committed:
more details, more tctl commands

sclevine committed:
Update 0169-auto-updates-linux-agents.md

sclevine committed:
scalability

sclevine committed:
df

sclevine committed:
content-length

sclevine committed:
cache init

sclevine committed:
binary

sclevine committed:
more rollout mechanism changes

sclevine committed:
scalability

sclevine committed:
more scalability

sclevine committed:
use 100kib pages for plan

sclevine committed:
Add RPCs, tweak API design

sclevine committed:
clarify wording

sclevine committed:
wording

sclevine committed:
Update rfd/0169-auto-updates-linux-agents.md

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

sclevine committed:
Update rfd/0169-auto-updates-linux-agents.md

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

sclevine committed:
linting

sclevine committed:
Move all RPCs into autoupdate/v1

sclevine committed:
Move groups to MVP

sclevine committed:
note about checksum

sclevine committed:
typos, consistency

sclevine committed:
clarify binary is teleport-update, package is teleport-ent-updater

sclevine committed:
switch from df to unix.Statfs

sclevine committed:
security feedback + naming adjustments

sclevine committed:
tweak rollout paging

sclevine committed:
tweak rollout paging again

sclevine committed:
feedback

sclevine committed:
adjust update.yaml to match implementation feedback

sclevine committed:
wip - new model

sclevine committed:
canaries

sclevine committed:
canary 2

hugoShaka committed:
describe state, transitions, and proxy response

sclevine committed:
rpcs

sclevine committed:
finish rpcs

sclevine committed:
minor tweaks

hugoShaka committed:
Add user stories

hugoShaka committed:
Put new requirements at the top + edit UX + add TODOs

hugoShaka committed:
Edition work

sclevine committed:
cleanup + swap phases 1 and 2

hugoShaka committed:
Move protobuf

hugoShaka committed:
Add installation scenarios

sclevine committed:
cleanup + move backpressure formulas

sclevine committed:
more cleanup

sclevine committed:
rename to unused number

sclevine committed:
fix title

sclevine committed:
more cleanup

sclevine committed:
correct inconsistencies

sclevine committed:
fix more inconsistencies

sclevine committed:
missing proxy flag

sclevine committed:
typo

zmb3 commented on rfd/0184-agent-auto-updates.md:
```diff
@@ -0,0 +1,1640 @@
+
+I can also install teleport using the package manager, then enroll the agent into AUs. See the section below:
+
+#### As a Teleport user I want to enroll my existing agent into AUs
```
Does this also cover existing agents using our current automatic upgrade system?

hugoShaka replied:
Yes, the UX should be the same. The updater might take extra steps to remove the old updater package from the system once everything is working.

zmb3 commented on rfd/0184-agent-auto-updates.md:
```diff
@@ -0,0 +1,1640 @@
+`teleport-update` will also handle `tbot` updates in the future, and it would be undesirable to distribute `teleport` with `tbot` just to enable automated updates.
+
+Finally, `teleport-update`'s API contract with the cluster must remain stable to ensure that outdated agent installations can always be recovered.
+The first version of `teleport-update` will need to work with Teleport v14 and all future versions of Teleport.
```
Even if we did have enough time, and managed to get this in, it would only be compatible with a very small percentage of v14 releases to date.

Probably easier to just support latest v15 and up.

sclevine replied:
Resolved during call: we can backport to v14 if necessary to migrate Cloud customers off v14.

hugoShaka committed:
Add CLI reference

sclevine commented on rfd/0184-agent-auto-updates.md:
````diff
@@ -0,0 +1,1640 @@
+$ teleport-update enable --proxy mytenant.teleport.sh --group staging
+```
+
+If the proxy address is not provided with `--proxy`, the current proxy address from `teleport.yaml` is used, if present.
````
It will have the same interface and default for specifying the data dir as the teleport binary:
https://github.com/gravitational/teleport/pull/46418/files#diff-d099d69928110c77b90c52fa55c9d973aa6070b7f1653c3eb6f6cd3c65352643R124

sclevine committed:
feedback

sclevine commented on rfd/0184-agent-auto-updates.md:
```diff
@@ -0,0 +1,1640 @@
+7. Download the desired Teleport tarball specified by `agent_version` and `server_edition`.
+8. Download and verify the checksum (tarball URL suffixed with `.sha256`).
+9. Extract the tarball to `/var/lib/teleport/versions/VERSION` and write the SHA to `/var/lib/teleport/versions/VERSION/sha256`.
+10. Update symlinks to point at the new version.
```
Added

sclevine commented on rfd/0184-agent-auto-updates.md:
````diff
@@ -0,0 +1,1640 @@
+  "agent_autoupdate": true,
+  "agent_update_jitter_seconds": 10
+}
+```
````
Truncated. Moved into an `auto_update` field as you suggested in the client tools PRs, and indicated that other fields are present.

sclevine commented on rfd/0184-agent-auto-updates.md:
```diff
@@ -0,0 +1,1640 @@
+        # Only valid for the backpressure strategy.
+        #  default: 20%
+        max_in_flight: 10-100%
+        # alert_after specifies the duration after which a cluster alert will be set if the group update has
+        # not completed.
+        #  default: 4
+        alert_after_hours: 1-8
```
Decided on cluster alerts during sync

sclevine committed:
alerts note

sclevine committed:
typos

sclevine commented on rfd/0184-agent-auto-updates.md:
````diff
@@ -0,0 +1,1640 @@
+```
+</details>
+
+I and the customer get an alert if the canary testing has not succeeded after an hour.
````
Clarified below: cluster alerts for customers, metrics-driven alerts for Cloud

sclevine committed:
Update rfd/0184-agent-auto-updates.md

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

sclevine committed:
clarify canary logic

sclevine commented on rfd/0184-agent-auto-updates.md:
```diff
@@ -0,0 +1,1640 @@
+
+#### Canary testing (phase 5)
+
+A group in `canary` state will get assigned canaries.
```
Clarified selection and update process.

sclevine commented on rfd/0184-agent-auto-updates.md:
```diff
@@ -0,0 +1,1640 @@
+        # wait_days specifies how many days to wait after the previous group finished before starting.
+        #  default: 0
+        wait_days: 0-1
+        # canary_count specifies the desired number of canaries to update before any other agents
+        # are updated.
+        #  default: 5
+        canary_count: 0-10
```
For now the selection will be random, max 10, with no control over which agents are selected within a group (clarified below)

hugoShaka commented on rfd/0184-agent-auto-updates.md:
```diff
@@ -1072,9 +1072,10 @@ update success criteria.
-The proxies will instruct those canaries to update now.
-During each reconciliation loop, the auth will lookup the instance heartbeat of each canary in the backend.
+A group in `canary` state will be randomly assigned `canary_count` canary agents.
+Auth servers will select those canaries by reading them from instance heartbeats and writing them to the `canaries` list in `agent_rollout_plan` status.
```
```suggestion
Auth servers will select those canaries by reading them from the auth instance inventory and writing them to the `canaries` list in `agent_rollout_plan` status.
```

sclevine commented on rfd/0184-agent-auto-updates.md:
```diff
@@ -0,0 +1,1640 @@
+If the entirely `teleport` binary were used instead, security scanners would match vulnerabilities all Teleport dependencies, so customers would have to handle rebuilding artifacts (e.g., AMIs) more often.
+Deploying these updates is often more disruptive than a soft restart of the agent triggered by the auto-updater.
+
+`teleport-update` will also handle `tbot` updates in the future, and it would be undesirable to distribute `teleport` with `tbot` just to enable automated updates.
```
I'm not sure if we'd want each binary to be individually self-updating. If all binaries have update commands, would `tbot update` be used to update teleport as well, or just tbot?

Seems cleaner to me to have a separate binary that is responsible for installing and updating all of the other binaries (including server-side tsh/tctl/tbot installs)?

sclevine committed:
Update rfd/0184-agent-auto-updates.md

Co-authored-by: Hugo Shaka <hugo.hervieux@goteleport.com>

sclevine committed:
Support for multiple installations / tarball

hugoShaka commented on rfd/0184-agent-auto-updates.md:
````diff
@@ -0,0 +1,1640 @@
+
+Alternatively
+```shell
+tctl autoupdate agent force-done dev
````
WDYT about:

```
tctl autoupdate agent mark-done dev
tctl autoupdate agent force-finish dev
```

hugoShaka committed:
Address reviewer's feedback

- Rephrase the UX section to not assume prior canary knowledge
- Explicit how the canaries are picked, the limitations, and potential
  improvements
- replace node with instance to avoid confusion between ssh nodes and
  generic teleport agent instances
- Explicit how the previous updater interacts with the new one
- More explicit names for command line args

hugoShaka committed:
agent_plan -> agent_rollout + reuse autoupdate_config

hugoShaka committed:
align tool version

sclevine committed:
Move package system dir

hugoShaka committed:
add time-based strategy

hugoShaka committed:
rename previous-must-succeed -> halt-on-failure

sclevine commented:
@russjones @zmb3 @rosstimothy unless there are any remaining questions, would be great to get this merged if you have a moment to approve 🙂
